src/bridge/cockpitdisksamples.c

```diff
@@ -153,3 +158,134 @@ cockpit_disk_samples (CockpitSamples *samples)
+                   guint64 read,
+                   guint64 write)
+{
+  cgroup_values_t *old = NULL;
+  if ((old = g_hash_table_lookup (table, cgroup)) != NULL)
+    {
+      // update values
+      old->read += read;
+      old->write += write;
+      g_free (cgroup);
+    }
+  else
+    {
+      // create new entry
+      cgroup_values_t *new = g_malloc0 (sizeof (cgroup_values_t));
+      new->read = read;
+      new->write = write;
+      g_hash_table_insert (table, cgroup, new);
+    }
+}
+
+static void
+get_process_io (const gchar *dir_name,
+                GHashTable *table)
+{
+  g_autofree gchar *path = g_strdup_printf ("/proc/%s/io", dir_name);
+  g_autofree gchar *contents = read_file (path);
+  if (!contents)
+    return;
+
+  guint64 read = 0, write = 0;
+  gchar **lines = g_strsplit (contents, "\n", -1);
+  for (guint n = 0; lines != NULL && lines[n] != NULL; n++)
+    {
+      gchar description[32] = {'\0'};
+      guint value;
+      const gchar *line = lines[n];
+      if (strlen (line) == 0)
+        continue;
+
+      if (sscanf (line, "%s %d", description, &value) != 2)
+        {
+          g_warning ("Error parsing line %d of file /proc/%s/io: `%s'", n, dir_name, line);
+          continue;
+        }
+
+      if (g_str_equal (description, "read_bytes:"))
+        read = value;
+      else if (g_str_equal (description, "write_bytes:"))
+        write = value;
+    }
+
+  g_strfreev (lines);
+
+  // get process cgroup
+  g_autofree gchar *cgroup_path = g_strdup_printf ("/proc/%s/cgroup", dir_name);
+  g_autofree gchar *cgroup = read_file (cgroup_path);
+  if (!cgroup)
+    return;
+  g_strchomp (cgroup);
+
+  table_add_values(table, g_steal_pointer (&cgroup), read, write);
+}
+
+static void
+sample_table_values (gpointer key,
+                   gpointer value,
+                   gpointer samples)
+{
+  // drop leadin '0::/' from cgroup name
+  cockpit_samples_sample ((CockpitSamples *)samples, "disk.cgroup.read", (gchar *)key + 4, ((cgroup_values_t *)value)->read);
+  cockpit_samples_sample ((CockpitSamples *)samples, "disk.cgroup.written", (gchar *)key + 4, ((cgroup_values_t *)value)->write);
+}
+
+void
+cockpit_cgroup_disk_usage (CockpitSamples *samples)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GDir) dir = g_dir_open ("/proc", 0, &error);
+
+  if (error)
+    {
+      g_warning ("Error when opening directory: %s", error->message);
+      return;
+    }
+
+  GHashTable *table = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+  g_autofree const gchar *dir_name = NULL;
+  while ((dir_name = g_dir_read_name (dir)) != NULL)
+    {
+      // directory name is PID of a process, PID is an unsigned int starting from 1
+      // when g_ascii_strtoull returns 0 it is not a PID directory
+      if (g_ascii_strtoull (dir_name, NULL, 10))
+        {
+          get_process_io (dir_name, table);
+        }
+    }
+
+  g_hash_table_foreach (table, sample_table_values, samples);
```

Comment:
It would be a lot clearer if you open-coded this with an iter:

```c
GHashTableIter iter;
gpointer key, value;
g_hash_table_iter_init (&iter, sample_table_values));
while (g_hash_table_iter_next (&iter, &key, &value))
  {
    const gchar *cgroup_name = key;
    const cgroup_values_t *values = value;

    // Skip ::0/
    cgroup_name += 4;

    ....etc
  }
```